### PR TITLE
feat: add pwa support

### PR DESCRIPTION
**Pull request by nezz0746**
- added "Launch as PWA" button on welcome page with install instructions package
- [x] Requesting feedback on UI of "Launch as PWA" button.
- added boilerplate code for web workers and ~~PUSH notifications~~ (REMOVED PUSH - Reduced scope for this PR)
- included manifest (icons, app name, etc)

### CODE REVIEW

File: next.config.js [hunk 1]
```diff
@@ -1,5 +1,25 @@
...
+    const interfaces = os.networkInterfaces()
+    for (const name of Object.keys(interfaces)) {
+        for (const iface of interfaces[name]) {
+            if (iface.family === 'IPv4' && !iface.internal) {
+                return iface.address
+            }
+        }
+    }
+    return null
+}
```
Comment (nezz0746) [inline next.config.js]: Used this to log local network url directly when launching the app & test pwa install procedures on mobile. Can remove.

**Comment by Hugo0** [comment]
Just testing out behaviour for now:

1. https://www.pwabuilder.com/reportcard?site=https://peanut-ui-git-feat-add-pwa-support-squirrellabs.vercel.app
Seems to think there's a few issues

2. On android it seems to install as a bookmark and not a pwa?
image

3. on linux/chrome the "install" button doesn't go away after it's been installed (But it is gone if i open it as a pwa). And if i click install again on chrome, then it gets stuck there and doesn't do anth new
image

File: custom.d.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+declare namespace JSX {
+    interface IntrinsicElements {
+        'pwa-install': React.DetailedHTMLProps<React.HTMLAttributes<HTMLElement>, HTMLElement>
+    }
+}
```
Comment (Hugo0) [inline custom.d.ts]: Q: why is this necessary? I have no idea - could we add a comment explaining why it's needed?

File: next.config.js (hunk 1 above)
Comment (Hugo0) [reply next.config.js]: I suggest we comment this code out then and add a comment on why it was there - in case we ever need it again for debugging

File: next.config.js [hunk 3]
```diff
@@ -107,49 +128,80 @@ const nextConfig = {
...
+    let conf = nextConfig
+    if (phase === PHASE_DEVELOPMENT_SERVER || phase === PHASE_PRODUCTION_BUILD) {
+        console.log('withSerwist')
+        const withSerwist = (await import('@serwist/next')).default({
+            swSrc: 'src/app/sw.ts',
+            swDest: 'public/sw.js',
+        })
+
+        conf = withSerwist(conf)
+    }
```
Comment (Hugo0) [inline next.config.js]: Q: what does this do :sob:

File: next.config.js [hunk 4]
```diff
@@ -107,49 +128,80 @@ const nextConfig = {
...
+        })
+    }
 
-        // Enables automatic instrumentation of Vercel Cron Monitors. (Does not yet work with App Router route handlers.)
-        // See the following for more information:
-        // https://docs.sentry.io/product/crons/
-        // https://vercel.com/docs/cron-jobs
-        automaticVercelMonitors: true,
-    })
+    return conf
```
Comment (Hugo0) [inline next.config.js:186]: weird diff but nth changed afaict

File: package.json [hunk 5]
```diff
@@ -7,6 +7,7 @@
     },
     "scripts": {
         "dev": "next dev",
+        "dev:https": "next dev --experimental-https",
```
Comment (Hugo0) [inline package.json:10]: q: what's this script for? Can we add a description in the README or somewhere?

File: package.json [hunk 6]
```diff
@@ -20,8 +21,10 @@
         "@dicebear/core": "^9.2.2",
         "@headlessui/react": "^1.7.19",
         "@headlessui/tailwindcss": "^0.2.1",
+        "@khmyznikov/pwa-install": "^0.4.6",
         "@safe-global/safe-apps-sdk": "^9.1.0",
         "@sentry/nextjs": "^8.30.0",
+        "@serwist/next": "^9.0.9",
```
Comment (Hugo0) [inline package.json:27]: Q: What's this new dependency?

File: package.json [hunk 7]
```diff
@@ -83,9 +87,11 @@
         "@types/three": "^0.157.2",
         "@types/uuid": "^9.0.8",
         "@types/validator": "^13.12.2",
+        "@types/web-app-manifest": "^1.0.8",
         "postcss": "^8.4.47",
         "prettier": "^3.3.3",
         "prettier-plugin-tailwindcss": "^0.5.14",
+        "serwist": "^9.0.9",
```
Comment (Hugo0) [inline package.json:94]: Q: dev dependency but also normal dep?

File: public/manifest.json [hunk 8]
```diff
@@ -1,14 +1,28 @@
...
+        {
+            "src": "/pwa/icon-192x192.png",
+            "sizes": "192x192",
+            "type": "image/png",
+            "purpose": "maskable"
+        },
+        {
+            "src": "/pwa/icon-512x512.png",
+            "sizes": "512x512",
+            "type": "image/png"
```
Comment (Hugo0) [inline public/manifest.json:15]: nice!

File: src/app/layout.tsx [hunk 9]
```diff
@@ -26,6 +28,15 @@ export default function RootLayout({ children }: { children: React.ReactNode })
                         <context.ContextProvider>
                             {children}
                             <CrispChat />
+                            <PWAInstaller
+                                name={manifest.short_name}
+                                description={manifest.description}
+                                icon={'/pwa/icon-512x512.png'}
+                                manual-apple="true"
+                                manual-chrome="true"
+                                // One space bc empty string is replaced with the default description
```
Comment (Hugo0) [inline src/app/layout.tsx:37]: what lol

File: src/app/sw.ts [hunk 10]
```diff
@@ -0,0 +1,59 @@
+import { defaultCache } from '@serwist/next/worker'
```
Comment (Hugo0) [inline src/app/sw.ts:4]: Q: i have no idea what thiis file does - can  we add a docstring explaining it?

File: src/components/Welcome/constants.ts [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+        classNameImg: 'h-6 w-6 sm:h-12 sm:w-12 rounded-lg',
+    },
+    {
+        icon: assets.NAZARE_LOGO_SQUARE,
+        link: 'https://www.nazare.io/',
+        classNameImg: 'rounded-full',
+    },
+]
+
+const faqs = [
```
Comment (Hugo0) [inline src/components/Welcome/constants.ts:42]: <3

File: src/components/Welcome/constants.ts [hunk 12]
```diff
@@ -0,0 +1,119 @@
+import * as assets from '@/assets'
+
+function classNames(...classes: any) {
+    return classes.filter(Boolean).join(' ')
+}
+
+const logoCloudLogos = [
```
Comment (Hugo0) [inline src/components/Welcome/constants.ts:7]: <3

File: src/constants/manifest.consts.ts [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+export const manifest: MetadataRoute.Manifest = {
+    name: 'Peanut Protocol',
+    short_name: 'Peanut',
+    description: 'Buttery smooth global money',
+    icons: [
+        {
+            src: '/pwa/icon-192x192.png',
+            sizes: '192x192',
+            type: 'image/png',
+        },
```
Comment (Hugo0) [inline src/constants/manifest.consts.ts:14]: Q: isn't this kinda duplicate with sw.ts?

File: src/hooks/useIsPWA.ts [hunk 14]
```diff
@@ -0,0 +1,20 @@
+import { useState, useEffect } from 'react';
```
Comment (Hugo0) [inline src/hooks/useIsPWA.ts:1]: tbr

**Review by Hugo0 — APPROVED** [review]

File: custom.d.ts (hunk 2 above)
Comment (panosfilianos) [reply custom.d.ts]: agree

File: next.config.js (hunk 1 above)
Comment (panosfilianos) [reply next.config.js]: thought: Ideally abstract this to a separate testing suite. Imo it's a bit weird to leave test code in prod components.
praise: gr8 thinking to test behavior this way

File: next.config.js [hunk 15]
```diff
@@ -107,49 +128,80 @@ const nextConfig = {
...
+            },
         ]
     },
 }
 
-const { withSentryConfig } = require('@sentry/nextjs')
+module.exports = async (phase) => {
+    let conf = nextConfig
+    if (phase === PHASE_DEVELOPMENT_SERVER || phase === PHASE_PRODUCTION_BUILD) {
+        console.log('withSerwist')
```
Comment (panosfilianos) [inline next.config.js]: thought: If this is properly QA'd, do we need this console.log?

File: next.config.js (hunk 3 above)
Comment (panosfilianos) [reply next.config.js]: agreed - and if this is needed for prod behavior, can we keep import at top of file?

File: next.config.js (hunk 4 above)
Comment (panosfilianos) [reply next.config.js:186]: I think this is tab changes

File: package.json (hunk 5 above)
Comment (panosfilianos) [reply package.json:10]: @Hugo0 this runs NextJS's server on https w/ a self signed cert.
thought: @nezz0746 is this necessarry to enable testing of installing the PWA?

File: package.json [hunk 16]
```diff
@@ -20,8 +21,10 @@
         "@dicebear/core": "^9.2.2",
         "@headlessui/react": "^1.7.19",
         "@headlessui/tailwindcss": "^0.2.1",
+        "@khmyznikov/pwa-install": "^0.4.6",
```
Comment (panosfilianos) [inline package.json:24]: thought: We should probably have a way of vetting new deps. This is built from a Microsoft eng afaik, but this can introduce security issues. Do we have any thoughts on this @Hugo0 ?

File: src/app/sw.ts (hunk 10 above)
Comment (panosfilianos) [reply src/app/sw.ts:4]: thought: additionally, if we've followed a guide/ tutorial to build this would be cool to link the resource.

File: src/components/PWAInstallComponent.tsx
Comment (panosfilianos) [inline src/components/PWAInstallComponent.tsx:1]: issue (non-blocking): I'd add this under a PWA folder in `components/`

File: src/components/PWAInstallComponent.tsx [hunk 17]
```diff
@@ -0,0 +1,82 @@
+'use client'
+
+import '@khmyznikov/pwa-install'
+import { PWAInstallElement } from '@khmyznikov/pwa-install'
+import { track } from '@vercel/analytics'
+import { useRef, useEffect, useCallback, useMemo } from 'react'
+
+export default function PWAInstaller({ ...props }) {
```
Comment (panosfilianos) [inline src/components/PWAInstallComponent.tsx]: issue (non-blocking): I'd explicitly define props here

File: src/components/PWAInstallComponent.tsx [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+    return <pwa-install id={`pwa-install`} {...props}></pwa-install>
+}
+
+export type PWAInstallerMethods = PWAInstallElement & {
```
Comment (panosfilianos) [inline src/components/PWAInstallComponent.tsx]: issue (non-blocking): I'd call this something like `PWAInstallerMethodsType`

File: next.config.js (hunk 3 above)
Comment (nezz0746) [reply next.config.js]: This is the service worker utility, it basically compiles your sw.ts file to javascript into the /public directory for runtime usage

File: custom.d.ts (hunk 2 above)
Comment (nezz0746) [reply custom.d.ts]: It's adding definition for the pwa-install component into the JSX namespace so that typescript doesn't kill us.
```ts
export default function PWAInstaller({ ...props }) {
    return <pwa-install id={`pwa-install`} {...props}></pwa-install>
}
```
But i'll move it to the global.d.ts actually, no need for a new separate declaration file

File: src/app/sw.ts (hunk 10 above)
Comment (nezz0746) [reply src/app/sw.ts:4]: Yes it's actually boilerplate code from serwist. DefaultCache is just a set or recommended cache strategies for the pwa. Adding link to docs ✅

File: src/constants/manifest.consts.ts (hunk 13 above)
Comment (nezz0746) [reply src/constants/manifest.consts.ts:14]: Do you mean manifest.ts ?

File: package.json (hunk 5 above)
Comment (nezz0746) [reply package.json:10]: There are some features of pwa and services workers that require being served over https, for example the issue I had with an event not firing right was only happening over https. Also testing out PUSH (removed for now) requires https. These cases come up here and there actually, just thought I'd make it available in the scripts.
Adding to README

File: next.config.js (hunk 3 above)
Comment (nezz0746) [reply next.config.js]: Also this module can't be required at the top (ES Module) and need to be imported dynamically
